ocaml/typing/env.ml

```diff
@@ -425,25 +425,33 @@ module IdTbl =
       | _ ->
           assert false
 
-    let add_lock mode next =
-      { current = Ident.empty; layer = Lock {mode; next} }
+    let add_lock lock next =
+      { current = Ident.empty; layer = Lock {lock; next} }
 
     let map f next =
       {
         current = Ident.empty;
         layer = Map {f; next}
       }
 
-    let rec find_same id tbl =
-      try Ident.find_same id tbl.current
+    let rec find_same_and_locks id tbl macc =
```

Comment:
I think this is a bad change. No caller of this function ever users the locks, and it would be suspicious if they did. The `find_same` functions are used primarily for looking up types and type-level components from things -- they should never care about locks. Their performance is also important since type lookup is one of the hottest parts of the compiler.

Reply:
The name nor the type of this function suggests that it's only used for looking up types. So I think it's more "correct" to not ignore locks.

In terms of performance: when this function is used for looking up types, the `macc` will stay as `[] : empty list`, which I think is an acceptable cost for safety. The list being `[]` is guaranteed by `empty` being uninhabited.

This is also similar to `find_name` below.

Reply:
It does somewhat tell you that, because it is looking for an `Ident.t` not a string. You are mostly only going to do that to lookup the definitions of types.

> the macc is statically guaranteed to stay at [] : empty list

That is only true if you are looking up something like `foo`, not if you are looking up `Foo.t` which is the more common case.

If you prefer, you can rename this function `find_same_without_locks` to make it clear. That is in fact clearer than `find_same_and_locks |> fst`. You can continue to expose it as `find_same` for the case where the type ensures that there are no locks.

Reply:
I agree with all what you said.